src/xo-app/pool/tab-patches.js

```diff
@@ -1,13 +1,160 @@
-import _ from 'intl'
+import ActionButton from 'action-button'
+import Component from 'base-component'
 import React from 'react'
+import SortedTable from 'sorted-table'
+import TabButton from 'tab-button'
+import _ from 'intl'
+import forEach from 'lodash/forEach'
+import isEmpty from 'lodash/isEmpty'
+import map from 'lodash/map'
 import { Container, Row, Col } from 'grid'
+import { connectStore } from 'utils'
+import {
+  createFilter,
+  createGetObjectsOfType
+} from 'selectors'
+import {
+  getHostMissingPatches,
+  installAllHostPatches
+} from 'xo'
+
+// ===================================================================
+
+const MISSING_PATCHES_COLUMNS = [
+  {
+    name: _('srHost'),
+    itemRenderer: host => host.name_label,
+    sortCriteria: host => host.name_label
+  },
+  {
+    name: _('hostDescription'),
+    itemRenderer: host => host.name_description,
+    sortCriteria: host => host.name_description
+  },
+  {
+    name: _('hostMissingPatches'),
+    itemRenderer: (host, missingPatches) => missingPatches[host.id],
+    sortCriteria: (host, missingPatches) => missingPatches[host.id]
+  },
+  {
+    name: _('patchUpdateButton'),
+    itemRenderer: host => (
+      <ActionButton
+        btnStyle='primary'
+        handler={installAllHostPatches}
+        handlerParam={host}
+        icon='host-patch-update'
+      />
+    )
+  }
+]
+
+@connectStore(() => {
+  const getHosts = createGetObjectsOfType('host').filter(
+    (_, props) => host => props.pool.id === host.$pool
+  )
+
+  return {
+    hosts: getHosts
+  }
+})
+export default class TabPatches extends Component {
+  constructor (props) {
+    super(props)
+    this.state.missingPatches = {}
+  }
+
+  _getHosts = createFilter(
+    () => this.props.hosts,
+    () => this.state.missingPatches,
+    [ (host, missingPatches) => missingPatches[host.id] ]
+  )
+
+  _refreshMissingPatches = async () => {
+    const missingPatches = {}
+    await Promise.all(
+      map(this.props.hosts, host => (
+        getHostMissingPatches(host).then(patches => {
+          missingPatches[host.id] = patches.length
+        })
+      ))
+    )
+    this.setState({ missingPatches })
+  }
+
+  _installAllMissingPatches = () => (
+    Promise.all(
+      map(this._getHosts(), installAllHostPatches)
+    )
+  )
+
+  componentWillMount () {
+    this._refreshMissingPatches()
+  }
+
+  componentWillReceiveProps (nextProps) {
+    forEach(nextProps.hosts, (host, push) => {
+      const { id } = host
+
+      if (this.state.missingPatches[id] !== undefined) {
+        return
+      }
+
+      this.setState({
+        missingPatches: {
+          ...this.state.missingPatches,
+          [id]: 0
+        }
+      })
+
+      getHostMissingPatches(host).then(patches => {
+        this.setState({
+          missingPatches: {
+            ...this.state.missingPatches,
+            [id]: patches.length
+          }
+        })
+      })
```

Comment:
Because ```componentWillReceiveProps``` is called if the hosts have changed. This function is not called if they are new patches, it's the big difference.

```componentWillReceiveProps``` is used here to receive patches from new hosts.

Note: ```getHostMissingPatches``` is an expensive function.